framework/include/controls/ParameterReceiver.h

```diff
@@ -0,0 +1,40 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// MOOSE includes
+#include "Control.h"
+
+// Forward declarations
+class Function;
+
+/**
+ * A Control object for receiving data from a master application transfer object.
+ */
+class ParameterReceiver : public Control
+{
+public:
+  static InputParameters validParams();
+
+  ParameterReceiver(const InputParameters & parameters);
+  virtual void execute() override;
+
+  /**
+   * Update the parameter names and associated values.
+   */
+  void transfer(const std::vector<std::string> & names, const std::vector<Real> & values);
```

Comment:
I don't particularly like this being public. @aeslaughter, what do you think? Is there a way we can create a base transfer object that is friends with this and have derived classes be able to call this correctly?

Reply:
Or maybe we can create an interface class.

Reply:
> I don't particularly like this being public. @aeslaughter, what do you think? Is there a way we can create a base transfer object that is friends with this and have derived classes be able to call this correctly?

@dschwen @lynnmunday and I had a meeting about this a few days ago, we decided it needed to be public so that custom transfers can use it. However, a better solution could be the @loganharbour write key trick and only give the key to Transfer objects. You can see how to do it here.

https://github.com/idaholab/moose/blob/e01d3f6df1bc9a848c00ce62996783361f99e2cd/framework/include/reporters/ReporterData.h#L47

Reply:
OK, what about this. The Transfer base class should contain a key class like I linked and create an instance at construction that is protected. Then any child Transfer class could use this key to call this method. We can then start using this key for other similar tasks were a generic Transfer object needs data access but we don't want to expose the world.

Reply:
If you want to do this, the inherited classes won't be able to construct it. You'd need to make the "key" object a protected member variable of the base.

Reply:
A control interface class might let me skip the receiver object and control parameters from the transfer object.

Reply:
I believe a a thought of a better option than using keys. We just create a function in a transfer base class, call it ParameterTransferBase. This class has a protected function that looks like this:
```
ParameterTransferBase::transferParameters(const std::vector<std::string> & names, const std::vector<Real> & values)
{
   ptr->transfer(names, values);
}
```
Derived transfers from this class can then use this function to transfer parameters. And the controller only needs to be friends with this function. This would work, right?

Reply:
Looks like SamplerParameterReceiver needs to get a new receiver for each sample, so maybe it needs to be:
```
void
ParameterTransferBase::transferParameters(ParameterReceiver & receiver, const std::vector<std::string> & names, const std::vector<Real> & values) const
{
   receiver.transfer(names, values);
}
```

Reply:
This sounds good. I"ll put it in.

Reply:
@zachmprince I have to add optimizationParameterTransfer as a friend class even though it's in a different app.  That makes the friend class awkward.

Reply:
I don't think you do. Only ParameterTransferBase needs to be a friend class. You can just call transferParameters in the derived class.

Reply:
You're right!  I"ll try again

Reply:
I added an interface.  I couldn't get the getReceiver function to work in the interface class, I have it in this branch: https://github.com/lynnmunday/moose/tree/paramterTransferWithGetReceiver 
but a lot of tests diff.  I think I'm mixing up receivers when I transfer.

Reply:
@zachmprince the tests pass.  What I put in could be a namespace with free functions and doesn't protect the transfer method